storage/jellyfish_merkle/src/restore/mod.rs

```diff
@@ -0,0 +1,411 @@
+
+            match self.partial_nodes[i].children[child_index] {
+                Some(ref child_info) => {
+                    // If the next node is an internal node, we just continue the loop with the
+                    // next nibble. Here we deal with the leaf case. We may need to insert multiple
+                    // internal nodes depending on the length of the common prefix of the existing
+                    // key and the new key.
+                    if let ChildInfo::Leaf { node } = child_info {
+                        assert_eq!(i, self.partial_nodes.len() - 1);
+
+                        let existing_leaf = node.clone();
+
+                        // The node at this position becomes an internal node. Since we may insert
+                        // more nodes at this position in the future, we do not know its hash yet.
+                        self.partial_nodes[i]
+                            .set_child(child_index, ChildInfo::Internal { hash: None });
+
+                        let common_prefix_len =
+                            existing_leaf.account_key().common_prefix_bits_len(new_key) / 4;
+
+                        // All these internal node will now have a single internal node child.
+                        for _ in i + 1..common_prefix_len {
+                            let visited_nibbles = nibbles.visited_nibbles().collect();
+                            let next_nibble = nibbles.next().expect("This nibble must exist.");
+                            let new_node_key = NodeKey::new(self.version, visited_nibbles);
+
+                            let mut internal_info = InternalInfo::new_uninitialized(new_node_key);
+                            internal_info.set_child(
+                                u8::from(next_nibble) as usize,
+                                ChildInfo::Internal { hash: None },
+                            );
+                            self.partial_nodes.push(internal_info);
+                        }
+
+                        // The last internal node will have two leaf node children.
+                        let visited_nibbles = nibbles.visited_nibbles().collect();
+                        let new_node_key = NodeKey::new(self.version, visited_nibbles);
+                        let mut internal_info = InternalInfo::new_uninitialized(new_node_key);
+
+                        // Next we put the existing leaf as a child of this internal node.
+                        let existing_child_index =
+                            existing_leaf.account_key().get_nibble(common_prefix_len);
+                        internal_info.set_child(
+                            existing_child_index as usize,
+                            ChildInfo::Leaf {
+                                node: existing_leaf,
+                            },
+                        );
+
+                        // Do not set the new child for now. Do it later after freezing the
+                        // previous leaf.
```

Comment:
The comment could be more informative - maybe explain that we cannot freeze the new child yet because it may end up colliding with the next restored account and creating another internal node.